_posts/2025-11-12-adaptive-refresh-for-resilient-segment-replication.md

```diff
@@ -0,0 +1,212 @@
+
+Lucene however, has a unique write once segmented architecture, where recently indexed documents are written into a self contained, immutable, Lucene segment. Once created, these segment files will never change again. While unusual, when compared to traditional databases, these immutable segments enable a lot of useful features like transactionality (a document is either fully indexed or not at all), a point in time view of the index that will never change, allowing for features like pagination, and multiple point in time snapshots - you can recover to an older snapshot by simply loading its corresponding segments. These properties enable another interesting form of replication called ‘*segment replication’*.
+
+With segment replication, a document is indexed only once on a primary indexing machine, the indexer. Immutable segments created in this process are then copied over the network to all the replicas. Whenever there are document updates or new documents get added, Lucene (on the indexer machine) indexes the changes and creates new segments, which are then copied over to all the replicas (searchers). On receiving the new segments (also called a replication payload or checkpoint), replicas atomically "refresh" their searchers to load these new segments. All subsequent search queries get served with the new index changes in place.
+
+With this approach, a document is indexed only once. We save precious computation resources on replicas, and instead leverage the network to copy (replicate) index changes across the entire fleet. By tracking replication checkpoints, we can make sure that each replica is on the same point-in-time view of the index, as opposed to document replication where each replica may index at its own pace. There's also the nice ability to rollback to a known older "good checkpoint" should some bad index changes make their way to production.
+
+
+## A typical segment replication setup
+
+In classic 'segment replicated' systems, indexers invoke a Lucene `commit` and create segments at fixed time intervals. Each commit is then treated as a replication checkpoint and copied over to replica instances. The frequency of creating checkpoints on indexers is pre-configured, based on the acceptable tolerance for propagating updates.  On receiving a new checkpoint, replicas compare it to their local state, fetch any missing files, and invoke a Lucene `refresh`, so that queries can see the updated index. 
```

Comment:
```suggestion
In classic "segment replicated" systems, indexers invoke a Lucene `commit` and create segments at fixed time intervals. Each commit is then treated as a replication checkpoint and copied over to replica instances. The frequency of creating checkpoints on indexers is preconfigured based on the acceptable tolerance for propagating updates. Upon receiving a new checkpoint, replicas compare it to their local state, fetch any missing files, and invoke a Lucene `refresh` so that queries can see the updated index. 
```